File: source/uCommonDevice.pas
unit uCommonDevice;

interface

uses
 SysUtils, uModbus;

type

  TCommonModbusDevice = class
  protected
    FModbus: TModBus;
    FIdent: TModbusIdentification;
  public
    constructor Create;
    destructor Destroy; override;

    function Open(Port, Speed, Adr: integer): TModbusError; overload;
    function Open(Port, Speed: integer): boolean; overload;
    function Open(Host: string; Port: integer): boolean; overload;
    function Close: boolean;

    function ReadDeviceIdentification(Addr: byte; var Ident: TModbusIdentification): TModbusError;
    function FindFirst: byte;
    function FindNext(FromID: byte): byte;

    function Connect: boolean;

    property DeviceInfo: TModbusIdentification read FIdent;
  end;

implementation

function TCommonModbusDevice.ReadDeviceIdentification(Addr: byte;
  var Ident: TModbusIdentification): TModbusError;
var
  s,
  dataout: ansistring;
  k: integer;
begin
  Ident.Company := '';
  Ident.Product := '';
  Ident.Version := '';

  for k := 0 to 3 do
  begin
    Result := FModbus.CallFunction(Addr, 43, 14, #$01 + ansichar(k), dataout);
    if Result <> merNone then exit;

    s := Copy(dataout, 11, ord(dataout[10]));
    if Pos(#$0, s) > 0 then
      s := Copy(s, 1, Pos(#$0, s) - 1);

    case k of
      0: Ident.Company := s;
      1: Ident.Product := s;
      2: Ident.Version := s;
      3: Ident.VendorURL := s;
      else
    end;

    if dataout[6] <> #$FF then break; // больше нечего тащить
  end;
end;

function TCommonModbusDevice.Close: boolean;
begin
  Result := false;
  try
    FModbus.Destroy;
    Result := true;
  except
  end;
  FModbus := nil;
end;

function TCommonModbusDevice.Connect: boolean;
begin
  if FModbus = nil then
  begin
    Result := false;
    exit;
  end;

  Result := FModbus.Connected;
end;

constructor TCommonModbusDevice.Create;
begin
  FModbus := nil;

  inherited;
end;

destructor TCommonModbusDevice.Destroy;
begin
  inherited;

  try
    FModbus.Free;
  except
  end;
end;

function TCommonModbusDevice.FindFirst: byte;
begin
  Result := FModbus.FindFirst;
end;

function TCommonModbusDevice.FindNext(FromID: byte): byte;
begin
  Result := FModbus.FindNext(FromID);
end;

function TCommonModbusDevice.Open(Port, Speed: integer): boolean;
begin
  Result := false;
  try
    FModbus := TModbus.Create(Port, Speed);
    if FModbus.Connected then
      Result := true;
  except
    FModbus := nil;
  end;
end;

function TCommonModbusDevice.Open(Port, Speed, Adr: integer): TModbusError;
begin
  FIdent.Company := '';
  FIdent.Product := '';
  FIdent.Version := '';

  FModbus := TModbus.Create(Port, Speed);
  try
    result := ReadDeviceIdentification(Adr, FIdent);
    if result <> merNone then
      FModbus.Destroy;
  except
    FModbus.Destroy;
  end;
end;

function TCommonModbusDevice.Open(Host: string; Port: integer): boolean;
begin
  Result := false;
  FIdent.Company := '';
  FIdent.Product := '';
  FIdent.Version := '';

  try
    FModbus := TModbus.Create(Host, Port);
    Result := ReadDeviceIdentification(200, FIdent) in [merNone, merTimeout];
  except
    FModbus := nil;
  end;
end;

end.

File: TurnstileCPP/uCommonDevice.cpp
#include "uCommonDevice.h"

TModbusError TCommonModbusDevice :: ReadDeviceIdentification(byte Addr)
{
	if(FIdent == NULL) return merCantOpenPort;
	return ReadDeviceIdentification(Addr, FIdent);
}

TModbusError TCommonModbusDevice :: ReadDeviceIdentification(byte Addr, TModbusIdentification * Ident)
{
    bytesArray s, dataout;
    bytesArray ttt;
    byte k;
    TModbusError err;
    int i;

    ttt.data[0] = 1;
    ttt.length = 2;

    for(k=0; k <=3; k++)
    {
        ttt.data[1] = k;
        err = FModbus->CallFunction(Addr, 43, 14, &ttt, &dataout);
        if(err != merNone )
        {
        	return err;
        }
        if(dataout.length >= 12)
        {
	        for(i=0; i<dataout.data[9] && i<dataout.length-12; i++)
	        {
	        	s.data[i] = dataout.data[i+10];
	        	if(!(s.data[i])) s.data[i] = 0x20; // replace zero with blank
	        }
	        s.data[i] = 0; 
	        s.length = i;
	        switch(k)
	        {
	            case 0:
	                strcpy(Ident->Company,(const char *)(s.data));
	                break;
	            case 1:
	                strcpy(Ident->Product,(const char *)(s.data));
	                break;
	            case 2:
	                strcpy(Ident->Version,(const char *)(s.data));
	                break;
	            case 3:
	                strcpy(Ident->VendorURL,(const char *)(s.data));
	                break;
	            default:
	                strcpy(Ident->Company,(const char *)s.data);
	                break;
	        }	        
        } 
        else
        {
        	s.data[0] = 0;
        	s.length = 0;
        	err = merWrongFromLength;
        	break;
        }         
        if(dataout.data[5]!= 0xFF)   break;  // No More Follows
    }
	return err;
}


void TCommonModbusDevice :: SetTimeOut(byte Value)
{
    FModbus->SetTimeOut(Value);
}


bool TCommonModbusDevice :: Close()
{
    if(FModbus != NULL)
    {
        delete FModbus;
        FModbus = NULL;
        return 1;
    }
    return 0;
}


bool TCommonModbusDevice :: Connect()
{
    if(FModbus == NULL)
    {
        return 0;
    }
    return FModbus->Connect();
}


TCommonModbusDevice :: TCommonModbusDevice()
{
    FModbus = NULL;
    FIdent = NULL;
    BlockMode = 1;  // Default mode for TCP Sockets is blocked mode
}


TCommonModbusDevice :: ~TCommonModbusDevice()
{
    if(FModbus!=NULL) delete FModbus;
    if(FIdent!=NULL) delete FIdent;
    FModbus = NULL;
    FIdent = NULL;
}


byte TCommonModbusDevice :: FindFirst()
{
    if(FModbus!=NULL) return FModbus->FindFirst();
    return 0;
}


byte TCommonModbusDevice :: FindNext(byte FromID)
{
    if(FModbus!=NULL) return FModbus->FindNext(FromID);
    return 0;
}


bool TCommonModbusDevice :: Open(int Port, int Speed)
{
    FIdent = new TModbusIdentification();
    FModbus = new TModbus(Port, Speed);
    if(FModbus != NULL && FIdent != NULL)
    {
        strcpy(FIdent->Company,"");
        strcpy(FIdent->Product,"");
        strcpy(FIdent->VendorURL,"");
        strcpy(FIdent->Version,"");
        return FModbus->Connect();
    }
    return 0;
}

bool TCommonModbusDevice :: Open(int Port, int Speed, int Adr)
{
    FModbus = NULL;
    FIdent = NULL;
    TModbusError err = merNone;
	
    FIdent = new TModbusIdentification();
    if(FIdent != NULL)
    {
        strcpy(FIdent->Company,"");
        strcpy(FIdent->Product,"");
        strcpy(FIdent->VendorURL,"");
        strcpy(FIdent->Version,"");
        FModbus = new TModbus(Port,Speed);

        if(FModbus != NULL)
        {
            err=ReadDeviceIdentification(Adr);
            if(err!=merNone && err != merTimeout)
            {
                if(FModbus!=NULL)
                {
                	delete FModbus;
                	FModbus = NULL;
                }
                if(FIdent!=NULL)
                {
                	delete FIdent;
                	FIdent = NULL;
                }
                return 0;
            }
            return 1; // OK
        }
        else
        {
        	if(FIdent!=NULL)
        	{
				delete FIdent;
				FIdent = NULL;
        	}
            return 0;
        }
    }
    return 0;
}


bool TCommonModbusDevice :: Open(char * Host, int Port)
{
	FIdent = new TModbusIdentification();
	FModbus = new TModbus(Host,Port,BlockMode);
    if(FModbus != NULL && FIdent != NULL)
    {
        strcpy(FIdent->Company,"");
        strcpy(FIdent->Product,"");
        strcpy(FIdent->VendorURL,"");
        strcpy(FIdent->Version,"");
        return FModbus->Connect();
    }
    return 0;
}

bool TCommonModbusDevice :: Open(char * Host, int Port, byte Adr)
{
    TModbusError err = merNone;
    FModbus = NULL;
    FIdent = NULL;
    FIdent = new TModbusIdentification();
    if(FIdent != NULL)
    {
        strcpy(FIdent->Company,"");
        strcpy(FIdent->Product,"");
        strcpy(FIdent->VendorURL,"");
        strcpy(FIdent->Version,"");
        FModbus = new TModbus(Host,Port);
        if(FModbus != NULL)
        {
            err = ReadDeviceIdentification(Adr);
            if(err!=merNone && err!=merTimeout)
            {
                if(FModbus != NULL)
                {
                	delete FModbus; 
                	FModbus = NULL;
                }
                if(FIdent != NULL)
                {
                	delete FIdent; 
                	FIdent = NULL;
                }                                
                return 0;
            }
            return 1;
        }
        else
        {
            if(FIdent != NULL)
            {
            	delete FIdent;
            	FIdent = NULL;
            }
            return 0;
        }
    }
    return 0;
}
